### ISSUE TITLE
Avoid repeat for optional fields with default factory?

### ISSUE THREAD

**Issue by Jerry-Ma** [pydantic/pydantic#9141]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Hi,
It came across to me that often I write models like this:
```
class MyModel(BaseModel):
    a: dict = Field(default_factory=dict)
    b: NestedModelThatHaveLongName = Field(default_factory=NestedModelThatHaveLongName)
```
I suspected that there would be an easy way to add some logic during the model building stage that we can do something like this:

```
class MyModel(BaseModel):
    a: dict = Field(default_factory=Self)
    b: NestedModeltahtHaveLongName = Field(default_factory=Self)
```
, where `Self` is a special constant (need a better name so it is not to be confused with `typing.Self`) that will be replaced with the actual type of the field.

The benefit of this is to remove unnecessary repeats and make code look more concise and focused.

What do you think?

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [X] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Viicos** [comment]
https://docs.pydantic.dev/latest/concepts/models/#fields-with-non-hashable-default-values (partially) covers your use case

**Comment by Jerry-Ma** [comment]
@Viicos

I read the section but I am not sure I am following. Could you show how I can leverage this feature to achieve (even partially) what I wanted, using my example in the OP as an example?

Are you suggesting the following?
```
class MyModel(BaseModel):
    a: dict = {}
    b: NestedModeltahtHaveLongName = NestedModeltahtHaveLongName()
```

As far as I am concerned, it is not the `Field(default_factory=)` part that I want to get rid of. It is the part that follows the`=` sign, i.e., the factory function, that I would like to eliminate with the proposed feature. The reasons is that, in most of the use cases, the type used to annotated the field serves as the `default_factory`, thus has to be repeated.

**Comment by Viicos** [comment]
Ah I see. This seems like a pretty niche case (as most of the time model names aren't this long), and I'd prefer having things explicit over a new `Self` (or similar, let's call it `DefaultFromAnnotation`) special constant. It is also unclear what should happen if:
- the type annotation used needs an argument:
```python
class SubModel(BaseModel):
    a: int

class Model(BaseModel):
    sub: SubModel = DefaultFromAnnotation
```
- the type annotation is an abstract type:
```python
from collections.abc import Sequence

class Model(BaseModel):
    sub: Sequence = DefaultFromAnnotation
```

These could be handled during model definition but would require some work.

But let's here what maintainers think of this!

**Comment by Jerry-Ma** [comment]
I really like the `DefaultFromAnnotation` construct. it would be even better if this can take some arguments just like `PrivateAttr` does:

```
class Model(BaseModel):
    sub: SubModel = DefaultFromAnnotation(description="The sub-model.")
```

As far as the more complex cases, I am thinking this `DefaultFromAnnotation` is more for types that are "default constructable". Any type that requires a non-trivial default_factory should have it specified manually.

**Comment by Viicos** [comment]
> As far as the more complex cases, I am thinking this `DefaultFromAnnotation` is more for types that are "default constructable". Any type that requires a non-trivial default_factory should have it specified manually.

It makes sense, I was referring to the fact that those kind of non-trivial/impossible cases (such as the ones I gave) would have to be caught during model creation, and that might be tricky to implement

**Comment by sydney-runkle** [comment]
Hey @Viicos and @Jerry-Ma,

I chatted with the team about this feature request this morning. For now, I don't think this fits into our roadmap. As @Viicos mentioned, I think there are a lot of edge cases with the usage that would get pretty complicated to handle, and explicitly using the type as the default works quite well in most cases. Is it the most concise, definitely not, but it gets the job done. I also agree that the usage is generally pretty niche, hence the lack of high priority here.

Closing for now as not planned. If there's significant demand for this in the future, we could reconsider implementing this!
